### PR TITLE
CI: move test_numpy_main to linux_meson

### PR DESCRIPTION
**Pull request by tupui**
Follow up of #16739

I am not sure about 2 things:

1. do we want the maintenance skip (`if: "github.repository == 'scipy/scipy' && !contains(github.ref, 'maintenance/') && !contains(github.base_ref, 'maintenance/')"`). I think this is a leftover as we have the `on: push: branches` specification.
2. do we really need `setuptools==59.8.0 wheel` in the 3.9 job??

### CODE REVIEW

**Comment by tupui** [comment]
@rgommers CI is green. LMK if there is anything to adapt here.

File: .github/workflows/linux_meson.yml [hunk 1]
```diff
@@ -51,9 +51,22 @@ jobs:
...
+      if: matrix.python-version == '3.11-dev'
+      run: |
+        python -m pip install git+https://github.com/numpy/numpy.git
+        python -m pip install ninja cython pytest pybind11 pytest-xdist pytest-timeout click rich-click doit pydevtool
+        python -m pip install git+https://github.com/serge-sans-paille/pythran.git
+        python -m pip install git+https://github.com/mesonbuild/meson.git
+        python -m pip install -r mypy_requirements.txt
+
+        # Packages that are only needed for their annotations
+        python -m pip install types-psutil pybind11 sphinx
```
Comment (rgommers) [inline .github/workflows/linux_meson.yml:68]: Better to move these two lines into the `name: Mypy` block for clarity now I think.

**Review by rgommers — COMMENTED** [review]
> 1. do we want the maintenance skip (`if: "github.repository == 'scipy/scipy' && !contains(github.ref, 'maintenance/') && !contains(github.base_ref, 'maintenance/')"`). I think this is a leftover as we have the `on: push: branches` specification.

We do need this. The first part is: "don't run on forks" (not a leftover, wanted in all jobs). The second part is "don't run the testing against main branches of dependencies in maintenance branches". That tends to fail and we disabled it on purpose.

There's one issue here now though: since you mixed this job with the `3.9` job, we cannot easily disable just one of these.

> > 2\. do we really need `setuptools==59.8.0 wheel` in the 3.9 job??

It looks like that is a leftover indeed, can you remove it?

**Comment by tupui** [comment]
> We do need this. The first part is: "don't run on forks" (not a leftover, wanted in all jobs). The second part is "don't run the testing against main branches of dependencies in maintenance branches". That tends to fail and we disabled it on purpose.
>There's one issue here now though: since you mixed this job with the 3.9 job, we cannot easily disable just one of these.

Yes I was referring to the second part only. I can just make a second conditional clause to skip if 3.11 & maintenance.

File: .github/workflows/linux_meson.yml [hunk 2]
```diff
@@ -26,11 +26,11 @@ jobs:
   test_meson:
     name: Meson build
     # If using act to run CI locally the github object does not exist and the usual skipping should not be enforced
-    if: "github.repository == 'scipy/scipy' || github.repository == ''"
+    if: "github.repository == 'scipy/scipy' || github.repository == '' && (matrix.python-version != '3.11-dev' && !contains(github.ref, 'maintenance/') && !contains(github.base_ref, 'maintenance/'))"
```
Comment (tupui) [inline .github/workflows/linux_meson.yml:29]:
```suggestion
    if: "github.repository == 'scipy/scipy' || github.repository == '' || (matrix.python-version != '3.11-dev' && !contains(github.ref, 'maintenance/') && !contains(github.base_ref, 'maintenance/'))"
```

File: .github/workflows/linux_meson.yml [hunk 3]
```diff
@@ -26,11 +26,11 @@ jobs:
   test_meson:
     name: Meson build
     # If using act to run CI locally the github object does not exist and the usual skipping should not be enforced
-    if: "github.repository == 'scipy/scipy' || github.repository == ''"
+    if: "github.repository == 'scipy/scipy' || github.repository == '' || (matrix.python-version != '3.11-dev' && !contains(github.ref, 'maintenance/') && !contains(github.base_ref, 'maintenance/'))"
```
[the remark this answers is not in the document]
Comment (tupui) [reply .github/workflows/linux_meson.yml:29]: Ooops yes!
```suggestion
    if: "github.repository == 'scipy/scipy' || github.repository == '' && !(matrix.python-version == '3.11-dev' && (contains(github.ref, 'maintenance/') || contains(github.base_ref, 'maintenance/')))"
```

**Comment by tupui** [comment]
After some googling, this happens because the matrix part cannot be used at that point. Apparently I could use exclude to do this. I will try on my fork if you prefer then.

**Comment by tupui** [comment]
Ok I think I got it with exclusion pattern and it's easier to read IMO. See https://github.com/tupui/scipy/pull/8 and https://github.com/tupui/scipy/pull/9 to see it in action.

**Comment by rgommers** [comment]
The Mypy step isn't erroring out when it should:
```
Installation OK
scipy/integrate/setup.py:4: error: Cannot find implementation or library stub for module named "scipy._build_utils"  [import]
scipy/integrate/setup.py:4: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
scipy/integrate/setup.py:9: error: Cannot find implementation or library stub for module named "scipy._build_utils.system_info"  [import]
scipy/stats/_hypotests.py:14: error: Cannot find implementation or library stub for module named "scipy.stats._hypotests_pythran"  [import]
Found 3 errors in 2 files (checked 783 source files)
```
That is already happening in `main` though, so let's ignore it here. I can reproduce it locally as well. I'll send a separate fix for that.

**Review by rgommers — APPROVED** [review]
LGTM, thanks @tupui!

**Pull request merged**